src/6model/reprs/MVMSpeshCandidate.c

```diff
@@ -0,0 +1,366 @@
+
+    /* Claim ownership of allocated memory assigned to the candidate. */
+    sg->cand = candidate;
+    MVM_spesh_graph_destroy(tc, sg);
+
+    /* Create a new candidate list and copy any existing ones. Free memory
+     * using the FSA safepoint mechanism. */
+    spesh = p->sf->body.spesh;
+    new_candidate_list = MVM_fixed_size_alloc(tc, tc->instance->fsa,
+        (spesh->body.num_spesh_candidates + 1) * sizeof(MVMSpeshCandidate *));
+    if (spesh->body.num_spesh_candidates) {
+        size_t orig_size = spesh->body.num_spesh_candidates * sizeof(MVMSpeshCandidate *);
+        memcpy(new_candidate_list, spesh->body.spesh_candidates, orig_size);
+        MVM_fixed_size_free_at_safepoint(tc, tc->instance->fsa, orig_size,
+            spesh->body.spesh_candidates);
+    }
+    MVM_ASSIGN_REF(tc, &(spesh->common.header), new_candidate_list[spesh->body.num_spesh_candidates], candidate);
+    spesh->body.spesh_candidates = new_candidate_list;
+
+    /* May now be referencing nursery objects, so barrier just in case. */
+    //if (spesh->common.header.flags2 & MVM_CF_SECOND_GEN)
+        //MVM_gc_write_barrier_hit(tc, (MVMCollectable *)spesh);
+
+    /* Regenerate the guards, and bump the candidate count only after they
+     * are installed. This means there is a period when we can read, in
+     * another thread, a candidate ahead of the count being updated. Since
+     * we set it up above, that's fine enough. The updating of the count
+     * *after* this, plus the barrier, is to make sure the guards are in
+     * place before the count is bumped, since OSR will watch the number
+     * of candidates to see if there's one for it to try and jump in to,
+     * and if the guards aren't in place first will see there is not, and
+     * not bother checking again. */
+    MVM_spesh_arg_guard_regenerate(tc, &(spesh->body.spesh_arg_guard),
+        spesh->body.spesh_candidates, spesh->body.num_spesh_candidates + 1);
+    MVM_barrier();
+    spesh->body.num_spesh_candidates++;
+
+    /* If we're logging, dump the updated arg guards also. */
+    if (MVM_spesh_debug_enabled(tc)) {
+        char *guard_dump = MVM_spesh_dump_arg_guard(tc, p->sf,
+                p->sf->body.spesh->body.spesh_arg_guard);
+        MVM_spesh_debug_printf(tc, "%s========\n\n", guard_dump);
+        fflush(tc->instance->spesh_log_fh);
+        MVM_free(guard_dump);
+    }
+
+#if MVM_GC_DEBUG
+    tc->in_spesh = 0;
+#endif
+}
+
+/* Discards existing candidates. Used when we instrument bytecode, and so
+ * need to ignore these ones from here on. */
+void MVM_spesh_candidate_discard_existing(MVMThreadContext *tc, MVMStaticFrame *sf) {
+    MVMStaticFrameSpesh *spesh = sf->body.spesh;
+    if (spesh) {
+        MVMuint32 num_candidates = spesh->body.num_spesh_candidates;
+        MVMuint32 i;
+        for (i = 0; i < num_candidates; i++)
+            spesh->body.spesh_candidates[i]->body.discarded = 1;
+        MVM_spesh_arg_guard_discard(tc, sf);
+    }
```

Comment:
We'll be able to do this rather more properly (actually discard them) once we get that mechanism working for deopt (for now though, I guess this stays as it is).